src/utils.rs

```diff
@@ -90,6 +91,14 @@ pub fn add_le(x: &mut [u8], y: &[u8]) -> Result<(), ()> {
     }
 }
 
+pub(crate) fn memzero_value<T>(v: T) {
```

Comment:
This should have `T: Copy` bound or call `ptr::drop_in_place`. Considering the intended use case, the `Copy` bound seems more appropriate.

Reply:
Whats the purpose in memzero-ing copied data?

Reply:
`T: Copy` does not mean that the data is copied, it means that it does not indirectly own any resources via e.g. pointer direction so that `Clone`ing it means to byte-for-byte copy it. If zeroing it out is to help with security, we need to avoid data types with indirections or follow all of the indirections to really zero out the relevant memory.

Reply:
But in practice it means that the struct will be memcpy’d into this function and then the copy will be securely erased. The compiler is free to give a reference, but in many cases a memcpy will happen which makes the whole function meaningless. You can verify this for many cases by looking at the generated assembly.

Reply:
> But in practice it means that the struct will be memcpy’d into this function and then the copy will be securely erased. 

This is true and was mentioned in both other threads here. But it does also affect the existing memzero-on-drop types as they can be moved - meaning `memcpy`ied around - just as well and `drop` will only zero out the last copy. Taking `self` by value only adds one more place where this problem manifests, but the problem itself is already present in the other wrapper types.

If we really want to fix this, then all sensitive data types have to be put into stable locations, e.g. using `Box` or `StableDeref` in general, and zeroing-on-drop must only work for those types and we must wrap them so user code will not accidentally copy/move out of them.

Reply:
> but they allow the user to make the choice if needed

I do not understand that. Which choice exactly do you mean? The compiler is free to move values around without user code explicitly requesting it and it can avoid to do so even if user code contains a move.

> however this function will always memcpy, and will thus be meaningless.

Not always, but yes the likelihood is high especially for small types. If we want to give something reliable here, we should restrict ourselves to something like `Box<T> where T: Copy` to get a stable pointer to zero-out before deallocation.

Reply:
> (and to be honest, most users don’t need to make that choice, since if someone can dump their running memory they have other more urgent problems)

I somewhat agree, but following this means that this API is not really worth the effort in the first place. Those very few users you need it can just unsafely call `utils::memzero` directly while having the highest chance to succeed using e.g. stable allocations.

Reply:
(I am not sure about the ergonomics of this, but `Pin<&mut T>` might also be useful choice.)

Reply:
Btw why so centered on not changing the api? Changes proposed in this PR is not destructive and allowed by rust semver convention (for next minor version)

Reply:
> I would be more ok with just adding the Drop implementation to Digest, like
we do with the other secret types. If it can contain secrets then we try to
erase it.

But generichash can also be used for non-confidential data and then forcefully erasing it might be a performance hit, even if minor

Reply:
> Btw why so centered on not changing the api? 

This is not my motivation. I'd just rather provide a general mechanism than do this again and again for every type. Especially since it really is optional for a digest. Also for `State`, there is still the problem of making the compiler take ownership but elide copies as much as possible.

> I would be more ok with just adding the Drop implementation to Digest, like
we do with the other secret types. If it can contain secrets then we try to
erase it.

This could imply a significant overhead as there are usually many more instances of digests than of e.g. keys. It would also imply that this should happen for SHA-256/512 as well. Making everyone pay that overhead for a best-effort mechanism is a questionable trade-off IMHO. Especially since the `Memzero` wrapper struct would it trivial to add this would modifying the `Digest` type at all.

Even for `State` something like
```
impl Memzero<T> {
  pub fn consume<F, R>(self, f: F) -> R where F: FnOnce(T) -> R {
    f(unsafe { ptr::read(self.0.as_ptr()) })
  }
}
```
would allow unmodified usage like `let digest = state.consume(State::finalize);` with less
 copying than the current proposal.

Reply:
> every type

I can't cone up with any type other than `Digest` and `generichash::State` for which this behavior would desired to be optional. All other types known to contain private data have automatically implemented `Drop`

Reply:
If you want to provide a way for users to erase their own structures, I think you should open separate PR as this is not really related to Digest.

Reply:
> All other types known to contain private data have automatically implemented Drop

What is different about `sha256::Digest` or `sha512::Digest` compared to `generichash::Digest`?

> If you want to provide a way for users to erase their own structures, I think you should open separate PR as this is not really related to Digest.

I think providing an API to zero out arbitrary `Copy` types is the most effective API to achieve this, especially since it implies that no modifications to `Digest` or `State` are necessary. I you want I can propose a separate MR for this, but I do think it would make this one unnecessary which is way I suggested modifying this one.

Reply:
#394

Reply:
> What is different

Nothing. They still are digests and just need the same api as blake2b digest. I missed them myself when implementing this, but I don't think it's worth to dance with state consuming instead of just extending digest api.

Reply:
> but I don't think it's worth to dance with state consuming instead of just extending digest api.

I had not though of that yet, but there is actually important ergonomic benefit to wrapper types: One cannot forget to call the zeroing `Drop´ implementation whereas it is quite possible to write a function that has some early returns, e.g. for error handling, that fail to call `drop_memzero`, `finalize_memzero` and friends. One would then probably put these calls into a scope guard and end up with something resembling the wrapper types in a more roundabout fashion.